### DISCUSSION TITLE
Login as arbitrary user when acting as admin

### DISCUSSION THREAD

**Discussion by ryami333** [directus/directus#19531] [Feature Requests]
### Summary

We would like a mechanism for logging in an arbitrary user using an admin's `access_token` (without knowing the users password). This would enable us to custom-build "magic link" login user-flows.

Currently, it's only possible to programmatically login a user with the `/auth/login` endpoint, but the user's password is a required field, so it must be known. Eg.

```ts
fetch("{URL}/auth/login", {
  method: "POST",
  body: JSON.stringify({ 
    email: "foo@bar.com",
    password: "xxxxxxxxx"
  })
});
```

We would like to be able to circumvent the `password` field by passing an Admin API token (using the typical authentication strategy: passing an `access_token` in the query-string, or by passing an `Authorization` header). Eg.

```ts
fetch("{URL}/auth/login", {
  method: "POST",
  body: JSON.stringify({ 
    email: "foo@bar.com",
  })
  headers: {
    Authorization: "Bearer xxxxxxxxxxxxxxx" // <- Admin access-token
  },
});
```

Alternatively, this could be a new endpoint altogether, so that the signature of the existing `/auth/login` endpoint would remain the same (password as a "required" field).

### Basic Example

Given above.

_No response_

### Motivation

This has at least two utilities, which would be of value to us.

### 1. Custom Authentication Mechanisms

We could implement custom user authentication strategies such as "magic link" (sometimes known as "passwordless") flows. For example, we could implement an API endpoint that looks like this:

```ts
// src/api/login/passwordless.ts

import { login } from '@directus/sdk';

export const GET = (request, response) => {
  // User arrives at this URL from a link we sent to them in an email.
  // The link contains a query-string like `?email=foo@bar.com&token=xxxxxxxx`.
  const { email, token } = request.params;

  const isValidToken = checkValidToken({ email, token });

  if (isValidToken) {
    const credentials = await fetch("{URL}/auth/login", {
      method: "POST",
      body: JSON.stringify({ 
        email: "foo@bar.com",
      })
      headers: {
        Authorization: "Bearer xxxxxxxxxxxxxxx" // <- Admin access-token
      },
    });
    
    return response.send({ credentials });
  }
  
  return response.setStatus(403).send({ error: "Forbidden" });
}
```

### 2. Impersonating Users

Some user-management softwares give you the ability to impersonate users, which is useful for the sake of development and testing. For example, [see Keycloak](https://www.keycloak.org/docs/latest/server_admin/#con-user-impersonation_server_administration_guide).

### Detailed Design

I think I've provided enough context and design information in the steps above, let me know if more is needed.

### Must Have:
- A mechanism for generating access tokens for arbitrary users as an admin.

### Should Have:
- N/A.

### Could Have:
- An environmental variable which enables/disables this behaviour

### Won't Have:
- N/A.

### Drawbacks

Some might consider this a security concern - I believe their concerns are misguided but this could be mitigated by making this configurable by opting-in with an environmental variable.

### Alternatives

Two different designs already considered: modification to existing `/auth/login` endpoint, or addition of new endpoint.

Impact of not doing this is that it's impossible to implement custom app-side auth strategies without it.

### Adoption Strategy

No breaking changes required if a new endpoint was introduced, no migrations required, as far as I can tell. But I'm sure maintainers will have a better feel for this than I do.

**Comment by rijkvanzanten** [comment]
Am I correct in believing this could be achieved with the syntax proposed in https://github.com/directus/directus/discussions/19085 ? 🙂

**Reply by ryami333** [reply]
I don't think so, no. That would only permit the second use-case I outlined in the motivations section. It still would not be possible to just get a user's access-token on their behalf.

**Reply by rijkvanzanten** [reply]
I suppose you could get the access token if you where to hit `/refresh` while impersonating that user 🤔

**Reply by ryami333** [reply]
I suppose. But in order to hit `/refresh` you need a refresh token, which you would need to obtain by first calling `/login` as admin, which you currently can't do with just an access-token, you'd need the admin _password_ (and email). So it's a bit of a circular issue.

**Reply by ryami333** [reply]
The issue could be reduced to: **it's not currently possible to login with an access_token**. This sounds like a silly/redundant assertion to make on the surface - it's only in edge-cases like this that it matters.

**Reply by rijkvanzanten** [reply]
> The issue could be reduced to: it's not currently possible to login with an access_token. This sounds like a silly/redundant assertion to make on the surface - it's only in edge-cases like this that it matters.

Which is by design! The refresh token is tied to a single session that's refreshed. If you can create new sessions with just the access token, you could create an infinite amount of new valid sessions, which would prevent you from force logging a user out 🤔

**Reply by ryami333** [reply]
Sure, that seems reasonable, just trying to illustrate why https://github.com/directus/directus/discussions/19085 doesn't resolve the concerns of this one.
